### PR TITLE
HIVE-24084 Push Aggregates thru joins in case it re-groups previously unique columns	

### PR DESCRIPTION
**Pull request by kgyrtkirk**
### What changes were proposed in this pull request?

* changes to cost estimations - fixed a few bugs
* changed Aggregate estimation to a different on HiveOnTezCostmodel
* in HiveAggregateJoinTransposeRule relaxed the comparision with the new plan to use <= instead <

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:293]: I think you suggested changing this... Maybe `isLe` if we do not introduce an additional aggregate on top?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:295]: Additionally, when we force triggering the transform, does it make sense to verify that we are not creating an aggregate on top (i.e., we end up with agg before join and after join?
That may narrow it down even further to a case when the pushdown should always be beneficial.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:352]: Can you use `JoinInfo.of` for this? It seems it does something very similar.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java
Comment (vineetgarg02) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java:106]: `rAverageSize` is based on input row count but `rowCount` is output row count. Is this intended or should average row size be computed based on output row count?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (vineetgarg02) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:319]: If the purpose of this method is to determine that given a set of columns are unique or not you can use `areColumnsUnique` as @jcamachor  suggested.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:341]: this method does a bit different thing - honestly I feeled like I'm in trouble when I've given this name to it :)

this method checks if the given columns contain an unique column somewhere in the covered joins; (this still sound fuzzy) so let's take an example

consider:
```
select c_id, sum(i_prize) from customer c join item i on(i.c_id=c.c_id)
```

* do an aggregate grouping by the column C_ID  ; and sum up something 
* below is a join which joins by C_ID
* asking wether C_ID  is a unique column on top of the join is false; but there is subtree in which C_ID is unique => so if we push the aggregate on that branch the aggregation will be a no-op

I think this case is not handled by `areColumnsUnique`

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:293]: yes; if we use `isLe` the current cost model which only takes rowcount into account will prefer the pushing aggregates further

there is another alternative to the `force` based approach: the rule can be configured to use the more advanced cost system - so that it could take cpu/io cost into account

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java:98]: maybe...I'm trying to catch the case when `inputRowCount >> outputRowCount`; we are also grouping - so it will not be a full sort at all ; I was using the above to achieve:
```
log(outputRowCount)*outputRowCount + inputRowCount*COST
```
the rational behind this is that it needs to really sort `oRC` and read `iRC` rows - this could be an underestimation...but `log(iRC)*iRC` was highly overestimating the cost

one alternative for the above could be to use:
```
log(outputRowCount) * inputRowCount
```

the rational behind this:
we will need to find the place for every input row; but we also know that the output will be at most `outputRowCount` - so it shouldn't take more time to find the place for the actual row than `log(outputRowCount)`

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java:106]: if we will be doing a 2 phase groupby: every mapper will do some grouping before it starts emitting; in case `iRC >> oRC` the mappers could eliminate a lot of rows ; and they will most likely utilize `O(oRC)` io

this is an underestimation ; I wanted to multiply it with the number of mappers - but I don't think that's known at this point....I can add a config key for a fixed multiplier.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:341]: Could you execute `areColumnsUnique` on the join input then? Wouldn't that simplify this logic?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java:98]: I think the problem is that we are trying to encapsulate here the algorithm selection too: The fact that we are grouping in each node before sorting the data (I think this is also somehow reflected in the `isLe` discussion above). However, that is not represented with precision by current model, since output rows is supposed to be the output of the final step in the aggregation.
Wrt read, there is also the IO part of the cost, I am trying to understand whether some of the cost representation that you are talking about is IO.
There is some more info about the original formulas that were used to compute this here: https://cwiki.apache.org/confluence/display/Hive/Cost-based+optimization+in+Hive
Can we split this into two patches and have the changes to the cost model on their own? This should also help to discuss this in more detail.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:341]: this method recursively checks that the above condition is satisfied or not - that's why it needs to call itself

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveOnTezCostModel.java]: sure; I'll open a separate ticket for the cost model changes

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:301]: I've added a config: `hive.transpose.aggr.join.unique` to enable/disable this feature

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java]: yes; I've explored using `areColumnsUnique` because it matches the usecase here - however for some tests it emitted some NPEs so I've gone back to the `getUniqueKeys` approach
I'll file a jira for `areColumnsUnique` when I know what's wrong with it..

File: ql/src/test/results/clientpositive/llap/constraints_optimization.q.out
Comment (jcamachor) [inline ql/src/test/results/clientpositive/llap/constraints_optimization.q.out:2634]: This is quite neat.

File: ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query23.q.out
Comment (jcamachor) [inline ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query23.q.out:161]: Cool!

File: ql/src/test/queries/clientpositive/tpch18.q
Comment (jcamachor) [inline ql/src/test/queries/clientpositive/tpch18.q:66]: Can we declare {{(L_ORDERKEY, L_LINENUMBER)}} as primary key? That way we will be able to verify whether HIVE-24087 is kicking in and removing the unnecessary join. 
We can also try the variant with {{L_ORDERKEY NOT NULL}} constraint.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:356]: As you go down recursively, you may start finding `HepRelVertex` as the rel node. I think you hit the `instanceof Project` below only for the first project because you create it using the builder.
That is why I think once you have gone through the first join, you will not hit another join; you could simply call `areColumnsUnique` in these `if` clauses, which could potentially uncover new cases.

I may be wrong though, I just wanted to leave you a final note to make sure it was clear what I meant.

**Review by jcamachor — COMMENTED** [review]
LGTM, left a couple of comments.

File: ql/src/test/queries/clientpositive/tpch18.q
Comment (kgyrtkirk) [reply ql/src/test/queries/clientpositive/tpch18.q:66]: I've added both constraints - it only removed the IS NOT NULL filter
it seems to me that 1 of the sum() is used as an output and the other is being used to filter by >300 - so both of them is being "used"

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:356]: That could be done; and I'm sure it was true in this - but this logic will work better if it could walk down as many joins as it could - we might have an aggregate on top in the meantime a bunch of joins under it...so I feel that it will be beneficial to retain it.
I feeled tempted to write a RelMd handler - however I don't think I could just introduce a new one easily.
RelShuttle doesn't look like a good match - I'll leave it as a set of `instanceof` calls for now.

I'll upload a new patch to see if digging deeper in the tree could do more or not.

**Review by jcamachor — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java:369]: OK. If it turns out there are many changes and it may need some time to be fixed, feel free to defer to follow-up JIRA and let's merge this one.

File: ql/src/test/queries/clientpositive/tpch18.q
Comment (jcamachor) [reply ql/src/test/queries/clientpositive/tpch18.q:69]: Thanks @kgyrtkirk .

This seems to need further exploration, we thought https://issues.apache.org/jira/browse/HIVE-24087 was going to help here. @vineetgarg02 , could you take a look at this once this patch is merged? Maybe the shape of the plan is slightly different to the one we anticipated.

**Pull request merged**
